WMass/python/postprocessing/examples/lepSFProducer.py

```diff
@@ -115,11 +115,162 @@ def analyze(self, event):
+            # just in case it was not loaded explicitly
+            if not self.hasLoadedHisto:
+                self.loadHist()
+                return self.getSF(pt,eta)
+            else: 
+                return 1.
+        else:
+            ieta = self.hist.GetXaxis().FindFixBin(eta+self.epsilon)
+            ipt =  self.hist.GetYaxis().FindFixBin(pt +self.epsilon)
+            # protect against underflow and overflow
+            return self.hist.GetBinContent(min(max(1,ieta),self.hist.GetNbinsX()),min(max(1,ipt),self.hist.GetNbinsY()))
+
+    def getSF_err(self,pt,eta):
+        if not self.hist:
+            #print "Warning in scaleFactorManager.getSF_err(): histogram not found, I will try to load it now. If not successfull, I will return 1."
+            # just in case it was not loaded explicitly
+            if not self.hasLoadedHisto:
+                self.loadHist()
+                return self.getSF(pt,eta)
+            else: 
+                return 1.
+        else:
+            ieta = self.hist.GetXaxis().FindFixBin(eta+self.epsilon)
+            ipt =  self.hist.GetYaxis().FindFixBin(pt +self.epsilon)
+            return self.hist.GetBinError(min(max(1,ieta),self.hist.GetNbinsX()),min(max(1,ipt),self.hist.GetNbinsY()))
+        
+#-----------------------------------------
+
+class lep2016SFProducer(Module):
+    def __init__(self):
+
+        self.mu_f = {"trigger"       :"smoothEfficiency_muons_trigger.root", 
+                     "identification":"smoothEfficiency_muons_full2016_ID.root", 
+                     "isolation"     :"smoothEfficiency_muons_full2016_ISO.root"
+                     }
+        self.el_f = {"trigger"       :"smoothEfficiency_electrons_trigger.root"
+                     }
+        self.filePath = "%s/src/CMGTools/WMass/python/postprocessing/data/leptonSF/new2016_madeSummer2018/" % os.environ['CMSSW_BASE']
+
+    def beginJob(self):
+        # create muon scale factor manager: pass file name and location, and then the name of histogram to read
+        # for muon ID, might also want to use "scaleFactorOriginal" which has the unsmoothed version of the scale factors
+        self.trgSF_manager_mu = scaleFactorManager(self.mu_f["trigger"],       self.filePath,"scaleFactor")
+        self.idSF_manager_mu  = scaleFactorManager(self.mu_f["identification"],self.filePath,"scaleFactor")  
+        self.isoSF_manager_mu = scaleFactorManager(self.mu_f["isolation"],     self.filePath,"scaleFactor")
+
+        # create electron scale factor manager        
+        self.trgSF_manager_el = scaleFactorManager(self.el_f["trigger"],       self.filePath,"scaleFactor")
+
+        # load histograms
+        self.trgSF_manager_mu.loadHist()
+        self.idSF_manager_mu.loadHist()
+        self.isoSF_manager_mu.loadHist()
+        self.trgSF_manager_el.loadHist()
+
+    def endJob(self):
+        pass
+    def beginFile(self, inputFile, outputFile, inputTree, wrappedOutputTree):
+        self.out = wrappedOutputTree
```

Comment:
Muons will have SF for:
1) Trigger 2) ID 3) ISO
Electrons will have SF for:
1) Trigger 2) RECO 3) full ID+ISO+Conversion rejection
Since they are 3 for both, I would sacrifice the name clarity for the same structure and better compression and call them sf1, sf2, sf3 for both